### PR TITLE
Split ConcreteModuleType into two types

### PR DESCRIPTION
**Pull request by suo**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #29826 Fix type sharing on loaded ScriptModules
* #29825 Re-unify module and interface in ConcreteModuleType
* **#29824 Split ConcreteModuleType into two types**

We have two distinct phases/uses for ConcreteModuleType:
1. We are building it up and using it to check whether we can
reuse JIT types. (RawConcreteModuleType)
2. We are using it to satisfy ModuleValue::attr queries.
(ConcreteModuleType)

These types share an underlying `ConcreteModuleTypeData` which 
actually stores the relevant info.

Previously they were the same type because I was lazy, but it's been the
source of a bug. So split them to formalize the differing invariants for
the two phases.

Differential Revision: [D18575010](https://our.internmc.facebook.com/intern/diff/D18575010)

### CODE REVIEW

File: torch/csrc/jit/script/concrete_module_type.h [hunk 1]
```diff
@@ -41,83 +43,24 @@ enum class IterableModuleKind { NONE, LIST, DICT };
...
-  // These getters are only here to return things as types that can be
-  // automatically converted by pybind.
-  std::unordered_map<std::string, py::object> getConstantsPy() const;
-  std::unordered_map<std::string, std::pair<TypePtr, bool>> getAttributesPy()
-      const;
-  std::vector<std::pair<std::string, TypePtr>> getModulesPy() const;
 
+// ConcreteModuleType has two phases.
+// 1. Creation: First we build it up, during the ScriptModule conversion
+// process. This is represented by RawConcreteModuleType.
```
Comment (zdevito) [inline torch/csrc/jit/script/concrete_module_type.h]: Before I found this comment, but after reading the diff I could not figure out which of the two types was which. Can we call RawConcreteModuleType ConcreteModuleTypeBuilder?

File: torch/csrc/jit/script/concrete_module_type.h [hunk 2]
```diff
@@ -256,11 +184,92 @@ class VISIBILITY_HIDDEN ConcreteModuleType {
...
-  ClassTypePtr jitType_ = nullptr;
   // NOTE: If you ever add any more state to this struct, you need to make sure
-  // operator== still makes sense! The only field that can be excluded from it
-  // is `jitType_`.
+  // operator== still makes sense!
+};
+
+// Represents a concrete type during in the process for construction. We use
+// this to decide whether we can share types between modules.
+class VISIBILITY_HIDDEN RawConcreteModuleType {
```
Comment (zdevito) [inline torch/csrc/jit/script/concrete_module_type.h]: If ConcreteModuleType was a friend of its builder, then we would not need ConcreteModuleTypeData and its data_ indirections. I don't think that changes its safety properties as long as the builder returns the built type by value to cause it to create a copy.

**Review by zdevito — APPROVED** [review]
Minor nits, but this otherwise looks good.

File: torch/csrc/jit/script/concrete_module_type.h (hunk 1 above)
Comment (suo) [reply torch/csrc/jit/script/concrete_module_type.h]: lol that was my first name for it 😛

**Pull request closed without merging**